crates/bevy_reflect/src/type_info.rs

```diff
@@ -0,0 +1,145 @@
+        TypeId::of::<T>() == self.1
+    }
+}
+
+/// Compile-time type information for various reflected types
+#[derive(Debug, Clone)]
+pub enum TypeInfo {
+    Struct(StructInfo),
+    TupleStruct(TupleStructInfo),
+    Tuple(TupleInfo),
+    List(ListInfo),
+    Map(MapInfo),
+    Value(ValueInfo),
+    /// Type information for "dynamic" types whose metadata can't be known at compile-time
+    ///
+    /// This includes structs like [`DynamicStruct`](crate::DynamicStruct) and [`DynamicList`](crate::DynamicList)
+    Dynamic(DynamicInfo),
+}
+
+impl TypeInfo {
+    /// The [`TypeIdentity`] of the reflected type
+    pub fn id(&self) -> &TypeIdentity {
+        match self {
+            Self::Struct(info) => info.id(),
+            Self::TupleStruct(info) => info.id(),
+            Self::Tuple(info) => info.id(),
+            Self::List(info) => info.id(),
+            Self::Map(info) => info.id(),
+            Self::Value(info) => info.id(),
+            Self::Dynamic(info) => info.id(),
+        }
+    }
+
+    /// The name of the reflected type
+    pub fn type_name(&self) -> &str {
+        match self {
+            Self::Struct(info) => info.id().type_name(),
+            Self::TupleStruct(info) => info.id().type_name(),
+            Self::Tuple(info) => info.id().type_name(),
+            Self::List(info) => info.id().type_name(),
+            Self::Map(info) => info.id().type_name(),
+            Self::Value(info) => info.id().type_name(),
+            Self::Dynamic(info) => info.id().type_name(),
+        }
+    }
+
+    /// The [`TypeId`] of the reflected type
+    pub fn type_id(&self) -> TypeId {
+        match self {
+            Self::Struct(info) => info.id().type_id(),
+            Self::TupleStruct(info) => info.id().type_id(),
+            Self::Tuple(info) => info.id().type_id(),
+            Self::List(info) => info.id().type_id(),
+            Self::Map(info) => info.id().type_id(),
+            Self::Value(info) => info.id().type_id(),
+            Self::Dynamic(info) => info.id().type_id(),
+        }
+    }
+}
+
+/// A container for compile-time info related to general value types, including primitives
+#[derive(Debug, Clone)]
+pub struct ValueInfo {
+    id: TypeIdentity,
+}
+
+impl ValueInfo {
+    pub fn new<T: Reflect + ?Sized>() -> Self {
+        Self {
+            id: TypeIdentity::of::<T>(),
+        }
+    }
+
+    /// The [`TypeIdentity`] of this value
+    pub fn id(&self) -> &TypeIdentity {
+        &self.id
+    }
+}
+
+/// A container for compile-time info related to general Bevy's _dynamic_ types, including primitives.
+///
+/// This is functionally the same as [`ValueInfo`], however, semantically it refers to dynamic
+/// types such as [`DynamicStruct`], [`DynamicTuple`], [`DynamicList`], etc.
+///
+/// [`DynamicStruct`]: crate::DynamicStruct
+/// [`DynamicTuple`]: crate::DynamicTuple
+/// [`DynamicList`]: crate::DynamicList
+#[derive(Debug, Clone)]
+pub struct DynamicInfo {
+    id: TypeIdentity,
```

Comment:
If all this does is wrap `TypeIdentity` do we need this anymore?

Reply:
This one is exactly the same as `TypeInfo::Value`, but I added it to mark a distinction between regular values and Bevy's dynamic ones. But I'm on the fence about keeping it. 

If we chose to get rid of it, we'd also have to decide if `Dynamic***` is considered a `TypeInfo::Value` or a `TypeInfo::Struct` or even just their respective type like `TypeInfo::List` for `DynamicList`. The reason we might want to go with the first option is that these types don't really contain any useful information since they're made to be dynamic. 

But I'd love to get others' thoughts on this.

Reply:
I guess the question is if we want to extend this with more metadata in the future, which is up in the air.

Visibility *could* be included here, but not sure if that is worth the trouble/memory.

Reply:
Yeah, though, I'm not sure _what_ metadata we'd store.

I'm thinking we might just want to get rid of it for now until that time comes. And now that I think about it, it probably makes sense to group them with their respective types:

* `DynamicStruct` → `TypeInfo::Struct`
* `DynamicTupleStruct` → `TypeInfo::TupleStruct`
* `DynamicTuple` → `TypeInfo::Tuple`
* `DynamicList` → `TypeInfo::List`
* `DynamicMap` → `TypeInfo::Map`

However, I'm not sure how to handle the key, value, and item types that the map and list info contain. Maybe they just return the `TypeIdentity` for the unit type? I'd rather not make those returns optional.

Otherwise, maybe we just consider all dynamic types `TypeInfo::Value`?

Reply:
Working a bit on improving serialization/deserialization for reflected types using `TypeInfo`, I think I found a proper use for `TypeInfo::Dynamic`.

Given the following RON:

```rust
// my_struct.ron
{
  "type": "my_crate::MyStruct",
  "value": {
    "dynamic_struct": {
      "foo": 123
    }
  }
}
```

How do we deserialize this (assuming `"dynamic_struct"` refers to a `DynamicStruct` type)? Depending on the deserializer, `foo` might be read as an `i32` or an `i64` or maybe a `usize`. This might be okay until a user tries to use it to patch a concrete instance, wherein it will fail if the types don't match.

By isolating `TypeInfo::Dynamic`, we can panic, warn, or do nothing when attempting to deserialize.

There might be other ways around this or it might not even be super important, but it's just one possible argument I found for keeping it as is.